jujugui/static/gui/src/app/components/search-results/search-results.js

```diff
@@ -21,100 +21,155 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
+     */
+    collapseSeries: function(entities) {
+
+      function seriesKey(series) {
+        // XXX kadams54, 2015-05-21: This series mapping needs to be updated
+        // with each release, at least until we can figure out a better way.
+        return {
+          precise: 12.04,
+          trusty:  14.04,
+          utopic:  14.10,
+          vivid:   15.04,
+          wily:    15.10
+        }[series.name];
+      }
+
+      function entityKey(entity) {
+        return [entity.name, entity.owner, entity.type, entity.promulgated];
+      }
+
+      var collapsedEntities = {},
+          orderedKeys = [];
+      for (var i = 0, l = entities.length; i < l; i++) {
+        var entity = entities[i],
+            key = entityKey(entity),
+            series = entity.series,
+            url = entity.url || '',
+            value = {name: series, url: url};
+        // If the key already exists, append the series to the existing list.
+        if (collapsedEntities[key]) {
+          var collapsed = collapsedEntities[key];
+          // Aggregate downloads across different series.
+          collapsed.downloads += entity.downloads || 0;
+          // Add series to the list.
+          collapsed.series.push(value);
+          // Ensure that the most recent series is first.
+          collapsed.series.sort(function(a, b) {
+            var k1 = seriesKey(a),
+                k2 = seriesKey(b);
+            // We want a reverse sort, so...
+            return k2 - k1;
+          });
+          // De-dupe the series array.
+          collapsed.series = collapsed.series.filter(function(s, pos, arry) {
+            return !pos || s.name != arry[pos - 1].name;
+          });
+          // And that its URL is used for the entity.
+          collapsed.url = collapsed.series[0].url;
+        } else {
+          // Convert all series attributes in the entities to lists.
+          if (series) {
+            entity.series = [value];
+          } else {
+            entity.series = [];
+          }
+          // Ensure downloads are present.
+          entity.downloads = entity.downloads || 0;
+          collapsedEntities[key] = entity;
+          // Save the key so we can preserve sort order.
+          orderedKeys.push(key);
+        }
+      }
+      // Now convert that object back to an ordered list and return it.
+      var returnedEntities = [];
+      for (var i = 0, l = orderedKeys.length; i < l; i++) {
+        var k = orderedKeys[i];
+        returnedEntities.push(collapsedEntities[k]);
+      }
+      return returnedEntities;
+    },
+
+    searchSuccess: function(rawResults) {
+      // Parse the raw results.
+      var results = rawResults.map(function(model) {
+        return model.toSearchResult(this.props.charmstore);
+      }, this);
+      results = this.collapseSeries(results);
+      // Split the results into promulgated and normal.
+      var promulgatedResults = results.filter(function(obj) {
```

Comment:
instead of two filter calls here you could have a conditional in the loop which would push to either a promulgated or normal results variables. Saving one loop